pkg/controller/cronjob/utils.go

```diff
@@ -147,6 +149,64 @@ func getRecentUnmetScheduleTimes(cj batchv1beta1.CronJob, now time.Time) ([]time
 	return starts, nil
 }
 
+// getRecentUnmetScheduleTimes2 gets a slice of times (from oldest to latest) that have passed when a Job should have started but did not.
+//
+// If there are too many (>100) unstarted times, just give up and return an empty slice.
+// If there were missed times prior to the last known start time, then those are not returned.
+func getRecentUnmetScheduleTimes2(cj batchv1beta1.CronJob, now time.Time, schedule cron.Schedule) ([]time.Time, error) {
+	starts := []time.Time{}
+
+	var earliestTime time.Time
+	if cj.Status.LastScheduleTime != nil {
+		earliestTime = cj.Status.LastScheduleTime.Time
+	} else {
+		// If none found, then this is either a recently created cronJob,
+		// or the active/completed info was somehow lost (contract for status
+		// in kubernetes says it may need to be recreated), or that we have
+		// started a job, but have not noticed it yet (distributed systems can
+		// have arbitrary delays).  In any case, use the creation time of the
+		// CronJob as last known start time.
+		earliestTime = cj.ObjectMeta.CreationTimestamp.Time
+	}
+	if cj.Spec.StartingDeadlineSeconds != nil {
+		// Controller is not going to schedule anything below this point
+		schedulingDeadline := now.Add(-time.Second * time.Duration(*cj.Spec.StartingDeadlineSeconds))
+
+		if schedulingDeadline.After(earliestTime) {
+			earliestTime = schedulingDeadline
+		}
+	}
+	if earliestTime.After(now) {
+		return []time.Time{}, nil
+	}
+
+	for t := schedule.Next(earliestTime); !t.After(now); t = schedule.Next(t) {
+		starts = append(starts, t)
+		// An object might miss several starts. For example, if
+		// controller gets wedged on friday at 5:01pm when everyone has
+		// gone home, and someone comes in on tuesday AM and discovers
+		// the problem and restarts the controller, then all the hourly
+		// jobs, more than 80 of them for one hourly cronJob, should
+		// all start running with no further intervention (if the cronJob
+		// allows concurrency and late starts).
+		//
+		// However, if there is a bug somewhere, or incorrect clock
+		// on controller's server or apiservers (for setting creationTimestamp)
+		// then there could be so many missed start times (it could be off
+		// by decades or more), that it would eat up all the CPU and memory
+		// of this controller. In that case, we want to not try to list
+		// all the missed start times.
+		//
+		// I've somewhat arbitrarily picked 100, as more than 80,
+		// but less than "lots".
+		if len(starts) > 100 {
```

Comment:
I'd drop this case, there are multiple issues about this limit. I'd throw event that we missed triggering a job (you can at most inform how many) and just continue with the last one.

Reply:
Also probably change the name of the function to relfect what it does. I doubt we need to do it similarly like in the old controller.

Reply:
+1 will handle this soon

Reply:
updated both the suggestions and related test cases. PTAL @soltysh